src/Castle.Facilities.AspNetCore/CastleWindsorRegistrationExtensions.cs

```diff
@@ -29,44 +29,185 @@ namespace Castle.Facilities.AspNetCore
 		/// <param name="container">Windsor container which activators call resolve against</param>
-		public static void AddCastleWindsor(this IServiceCollection services, IWindsorContainer container)
+		/// <param name="configure">Configuration options for controlling registration and lifestyles of controllers, tagHelpers and viewComponents</param>
+		public static void AddCastleWindsor(this IServiceCollection services, IWindsorContainer container, Action<CastleWindsorRegistrationOptions> configure = null)
 		{
-			services.AddRequestScopingMiddleware(container.BeginScope);
-			services.AddCustomControllerActivation(container.Resolve);
-			services.AddCustomTagHelperActivation(container.Resolve);
-			services.AddCustomViewComponentActivation(container.Resolve);
-			container.Kernel.Resolver.AddSubResolver(new LoggerDependencyResolver(services));
-			container.Kernel.Resolver.AddSubResolver(new FrameworkConfigurationDependencyResolver(services));
+			AddFrameworkResolversToWindsor(services, container);
+			InstallFrameworkIntegration(services, container);
+
+			var options = new CastleWindsorRegistrationOptions(container, services);
+			(configure ?? ((o) => { }))(options);
+
+			AddApplicationComponentsToWindsor(container, options);
 		}
 
+
 		/// <summary>
-		/// Use this to register all Controllers, ViewComponents and TagHelpers
+		/// For making types available to the IServiceCollection using 'late bound' factories resolved via Windsor. This makes things like the @Inject directive in Razor work.
 		/// </summary>
-		/// <typeparam name="TTypeAssembly">Type from assembly to scan and register ASP.NET Core components</typeparam>
-		/// <param name="app">Application builder retained as extension</param>
-		/// <param name="container">Windsor container to register framework types in</param>
-		public static void UseCastleWindsor<TTypeAssembly>(this IApplicationBuilder app, IWindsorContainer container)
+		/// <param name="registration">The component registration that gets copied across to the IServiceCollection</param>
+		/// <param name="serviceType">The type of the service</param>
+		/// <param name="container">Windsor container which 'activators'/'cross wired' components resolve from</param>
+		/// <param name="serviceCollection">ASP.NET Core service collection from Microsoft.Extensions.DependencyInjection</param>
+		/// <param name="lifestyleType">The lifestyle type that needs to be copied across, only Singleton, Scoped and Transient are supported</param>
+		public static ComponentRegistration<TService> CrossWired<TService>(this ComponentRegistration<TService> registration, Type serviceType, IWindsorContainer container, IServiceCollection serviceCollection, LifestyleType lifestyleType = LifestyleType.Scoped) where TService : class
 		{
-			var assembly = typeof(TTypeAssembly).Assembly;
-			UseCastleWindsor(container, assembly);
+			registration.LifeStyle.Is(lifestyleType);
```

Comment:
The comment says only 3 lifestyle types are supported, should this check nothing else is passed in?

Reply:
Yes we should design by contract. Do you know of any way we could integrate the other types? I personally have never needed them in a web scenario.

Reply:
> Do you know of any way we could integrate the other types?

Someone could create any sort of custom lifestyle type, not sure you'll be able to handle that with this facility, probably just worth throwing on unsupported lifestyle until we need them and have a proper use case.

Reply:
Done